validation: distinguish between same tx and same-nonwitness-data tx in mempool

glozow opened the pull request:
On master, if you submit a transaction with the same txid but different witness to the mempool, it thinks the transactions are the same. Users submitting through `BroadcastTransaction()` (i.e. `sendrawtransaction` or the wallet) don't get notified that there's a different transaction in the mempool, although it doesn't crash. Users submitting through `testmempoolaccept()` will get a "txn-already-in-mempool" error.

This PR simply distinguishes between `txn-already-in-mempool` and `txn-same-nonwitness-data-in-mempool`, without handling them differently: `sendrawtransaction` still will not throw, but `testmempoolaccept` will give you a different error.

I believe the intention of #19645 is to allow full swaps of transactions that have different witnesses but identical nonwitness data. Returning a different error message + adding a test was suggested: https://github.com/bitcoin/bitcoin/pull/19645#issuecomment-705109193 so this is that PR.

glozow committed:
[test] submit same txid different wtxid as mempool tx

Co-authored-by: Antoine Riard <ariard@student.42.fr>
Co-authored-by: Antoine Riard <antoine.riard@gmail.com>

glozow committed:
[validation] distinguish same txid different wtxid in mempool

Changes behavior.

jnewbery commented on src/validation.cpp:
```diff
@@ -611,9 +611,13 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
         return state.Invalid(TxValidationResult::TX_CONFLICT, "txn-already-in-mempool");
+    } else if (m_pool.exists(GenTxid(false, tx.GetHash()))) {
+        // Transaction with the same nonwitness data but different witnes (same txid, different
+        // wtxid) already exists in the mempool. TODO: allow replacements
```
TODOs in comments are generally discouraged. If someone wants to implement witness replacement, they don't need a TODO to tell them to do that.

glozow replied:
Gotcha, fixed.

jnewbery commented on test/functional/mempool_wtxid.py:
```diff
@@ -0,0 +1,123 @@
+        node = self.nodes[0]
+
+        self.log.info('Start with empty mempool, and 200 blocks')
+        self.privkeys = [node.get_deterministic_priv_key().key]
+        self.address = node.get_deterministic_priv_key().address
+        self.coins = []
```
No need for these to be member variables:

```suggestion
        privkeys = [node.get_deterministic_priv_key().key]
        address = node.get_deterministic_priv_key().address
        coins = []
```

glozow replied:
Done - got rid of some of it entirely since not needed for `generate()`

jnewbery commented on test/functional/mempool_wtxid.py:
```diff
@@ -0,0 +1,123 @@
+        self.address = node.get_deterministic_priv_key().address
+        self.coins = []
+        # The last 100 coinbase transactions are premature
+        for b in node.generatetoaddress(110, self.address)[:-100]:
```
I think you can just call `node.generate()` and it'll generate to the node's deterministic address.

glozow replied:
Ohhh I didn't know that, I've been using `generatetoaddress` this whole time :scream:

glozow replied:
Done

jnewbery commented on test/functional/mempool_wtxid.py:
```diff
@@ -0,0 +1,123 @@
+        assert_equal(node.getrawmempool(True)[txid_submitted]['wtxid'], child_one_wtxid)
+
+        # testmempoolaccept reports the "already in mempool" error
+        assert_equal(node.testmempoolaccept([ToHex(child_one)]),
```
can you break this into multiple lines like the `assert_equal()` below?

glozow replied:
Done

jnewbery reviewed (commented):
Code review ACK 012341f779f4b82a757f30ef36810d8f8a1975dc

A few style comments inline. Feel free to ignore.

glozow commented:
CC @ariard - The same-txid-different-wtxid issue was initially pointed out in #19645 and the test commit is converted from https://github.com/bitcoin/bitcoin/pull/19645/commits/d86e7a17081af04ceabe93b15a87cf295a9e9d84

ariard commented on src/node/transaction.cpp:
```diff
@@ -50,7 +50,7 @@ TransactionError BroadcastTransaction(NodeContext& node, const CTransactionRef t
         if (!existingCoin.IsSpent()) return TransactionError::ALREADY_IN_CHAIN;
     }
-    if (!node.mempool->exists(hashTx)) {
+    if (!node.mempool->exists(GenTxid(false, tx->GetHash()))) {
```
Is this change altering behavior of `BroadcastTransaction` ? Otherwise maybe better to just rename `hashTx` to `txid`, update comment to "Txid is not already in the mempool. Note if the submitted candidate differs by wtxid from an already present transaction, it's for now rejected", and make it its own commit as it's only a documenting change ?

glozow replied:
Good point, my intention is merely to highlight/document what the current behavior is. I've now marked the commits with "changes / doesn't change behavior" and switched this to just be a comment explaining the behavior.

ariard commented on test/functional/mempool_wtxid.py:
```diff
@@ -0,0 +1,123 @@
+"""
+Test mempool acceptance in case of an already known transaction
+with identical non-witness data different witness.
+"""
```
nit: I think the test could be called "mempool_wtxid_acceptance" because this is what it does exercising our mempool logic w.r.t to accepting candidates identified by wtxid imo ?

glozow replied:
Done!

ariard commented on src/validation.cpp:
```diff
@@ -611,9 +611,13 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
+    } else if (m_pool.exists(GenTxid(false, tx.GetHash()))) {
+        // Transaction with the same nonwitness data but different witnes (same txid, different
+        // wtxid) already exists in the mempool. TODO: allow replacements
+        return state.Invalid(TxValidationResult::TX_CONFLICT, "txn-same-nonwitness-data-in-mempool");
```
What do you think about "mutant-replacement-disallowed" as a rejection message ? Where a mutant is defined as a transaction sharing an equivalent txid of an already in-mempool transaction but offering differing witnesses. 

It's quite kind of a glossary point but i don't think we have yet an unified term across the codebase to designate same-txid-diff-wtxid ? Maybe someone has a better naming proposal, but it sounds less verbose rather than duplicating "same nonwitness data but different witness (same txid, different wtxid)" everywhere in the code paths where this subtlety is present.

glozow replied:
I don't think a user would find "mutant-replacement-disallowed" more helpful than "txn-same-nonwitness-data-in-mempool." It would require them knowing what mutant replacement is (which doesn't exist yet), which requires knowing what a mutant is (which is not widely defined). It's confusing because it can imply that mutant replacement _is_ allowed somehwere else...

If/when the replacement policy is implemented, that would be an excellent time to give it a name. But at this point, what's relevant to users is "a different transaction with the same nonwitness data is already in the mempool."

ariard replied:
Fair point, it's quite a chicken-and-egg issue, either you pollute the codebases with multiple times verbose variable names like "same-txid-different-wtxid", which are less unambiguous or you find one name, defined somehwhere in the codebase, and hope to reviewers to refer it during discussion.

Yes it can be added latter for sure, once we add mutant replacement.

ariard commented on src/validation.h:
```diff
@@ -181,12 +181,15 @@ struct MempoolAcceptResult {
     /** Raw base fees in satoshis. */
     const std::optional<CAmount> m_base_fees;
+    /** Size in virtual bytes. */
+    const std::optional<int64_t> m_vsize;
```
Do you plan to use vbytes as an identifier to dissociate same-txid-different-wtxid transaction in `AcceptSingleTransaction` consumers ?

I'm not sure it's absent of collisions, two mutants can have the same vbytes but yet different witnesses ? E.g, a P2WSH 1-of-2 spent by either Alice signature or Bob signature

glozow replied:
Nope. See https://github.com/bitcoin/bitcoin/pull/22252/commits/188cab0af2c5ee646f1467bd0a300d361ffd1fed - we need the vsize in the result because otherwise the RPC code doesn't know it without looking into the mempool (which would require holding the mempool lock).

ariard replied:
I don't think the commit you're pointing to me is a unittest, exercising this logic with a collision on vbytes ? Do you think we can't have vbytes collision among "same-txid-different-wtxids" candidates ?

> we need the vsize in the result because otherwise the RPC code doesn't know it without looking into the mempool (which would require holding the mempool lock).

Sorry, I don't understand, what the RPC code is searching to know, the vbyte's size or an ambiguous identifer between "same-txid-different-wtxids" candidates ?

ariard reviewed (commented):
So I'm Concept ACK on returning a different `testmempoolaccept` error message when we hit a same-txid-different-wtxid mempool candidate but I don't think this is necessary for current package mempool accept project ? Enabling witness replacement is a project which is worthy on its own imho, as Taproot make it more probable to have concurrent broadcasts with "feerate-divergent" witnesses.

Note, to allow same-txid-different-wtxid we still need to rework RBF rules, as the rule 5 enforces a replacement penalty which imply an absolute fee increases IIRC. However, an absolute fee is implicitly committed as part of non-witness data (sub outpoints's amount/outputs's value).

jnewbery commented on src/node/transaction.cpp:
```diff
@@ -51,7 +51,11 @@ TransactionError BroadcastTransaction(NodeContext& node, const CTransactionRef t
     }
     if (!node.mempool->exists(hashTx)) {
-        // Transaction is not already in the mempool.
+        // A transaction with the same txid is not already in the mempool. Note that it is still
+        // possible for a transaction with same nonwitness data but different witness to be in the
+        // mempool - in that case, we treat them as the same transaction and don't try to submit tx
+        // to mempool. This logic must be changed if we allow replacements of
+        // same-txid-different-wtxid transactions in the future.
```
I find this comment quite confusing. I initially read the "Note that it is still possible for a transaction with same nonwitness data but different witness to be in the mempool." as meaning "it is possible for this condition to be true in this code branch".

Does this seem better to you:

```diff
-    if (!node.mempool->exists(hashTx)) {
-        // A transaction with the same txid is not already in the mempool. Note that it is still
-        // possible for a transaction with same nonwitness data but different witness to be in the
-        // mempool - in that case, we treat them as the same transaction and don't try to submit tx
-        // to mempool. This logic must be changed if we allow replacements of
-        // same-txid-different-wtxid transactions in the future.
+    if (node.mempool->exists(hashTx)) {
+        // There's already a transaction in the mempool with this txid. Don't try to submit
+        // this transaction to the mempool, but do attempt to reannounce it if relay=true.
+        // The mempool transaction may have the same or different witness (and wtxid) as this
+        // transaction.
+    } else {
```

jnewbery commented on test/functional/mempool_accept_wtxid.py:
```diff
@@ -0,0 +1,126 @@
+
+        self.log.info('Start with empty mempool, and 200 blocks')
+        privkeys = [node.get_deterministic_priv_key().key]
+        coins = []
+        # The last 100 coinbase transactions are premature
+        for b in node.generate(110)[:-100]:
+            coinbase = node.getblock(blockhash=b, verbosity=2)["tx"][0]
+            coins.append({
+                "txid": coinbase["txid"],
+                "amount": coinbase["vout"][0]["value"],
+                "scriptPubKey": coinbase["vout"][0]["scriptPubKey"],
+            })
+        assert_equal(node.getmempoolinfo()['size'], 0)
+
+        txid = coins.pop()["txid"]
```
You don't need a `coins` list or all this logic. Just generate 101 blocks, then get the txid of the first transaction in the tip block.

jnewbery commented:
ACK e6ba22edbc13f29d53f2dc0fcdcff49df879f597

ariard commented:
@glozow 

You know i'm quite bothered by you reaching offline, asking me to re-use my work with #19645 and the well-foundness #22252, of which I said it wasn't really required for a v0.1 package-relay and that we should be [careful](https://github.com/bitcoin/bitcoin/pull/22252#issuecomment-861930852) before considering this change.

I still appreciate the energy and dedication you're pouring in this serie of work. That said next time if you ask questions on any work related to better support for L2s, ping me on #bitcoin-core-dev. Anyone will be able to learn from the discussion, and it would avoid me to repeat twice my arguments :/

glozow commented:
@ariard We discussed this [8 months ago on github, when you posted that you would open the PR to make this change](https://github.com/bitcoin/bitcoin/pull/19645#issuecomment-705109193). I felt that there was a bit of urgency which is why I wanted to PR it, and messaged you privately to make sure we wouldn't be doing redundant work.

glozow commented:
I've updated the description and removed commits that aren't strictly about same-txid-different-wtxid. Hopefully #22252 no longer needs to be considered for this PR!

jnewbery commented:
Code review ACK ea678026a7ca41659a09cab9ab5b989eed7d3bf9

darosior reviewed (approved):
utACK ea678026a7ca41659a09cab9ab5b989eed7d3bf9

naumenkogs commented on src/validation.cpp:
```diff
@@ -611,9 +611,13 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
+        // Exact transaction already exists in the mempool.
         return state.Invalid(TxValidationResult::TX_CONFLICT, "txn-already-in-mempool");
+    } else if (m_pool.exists(GenTxid(false, tx.GetHash()))) {
+        // Transaction with the same nonwitness data but different witnes (same txid, different
```
typo: "witnes"

glozow replied:
Fixed

naumenkogs commented:
The last commit has Antoine emails twice: 
```
Co-authored-by: Antoine Riard <ariard@student.42.fr>
Co-authored-by: Antoine Riard <antoine.riard@gmail.com>
```

Not that I care much, just pointing out it might be something odd.

---------------------------

At a high level, I'm not getting lost with the justification of this in terms of [19645](https://github.com/bitcoin/bitcoin/pull/19645#issuecomment-705109193).

If the plan is to have replacement of different-witness in the future, then why adding this error in the first place here? Do we need this in the meanwhile, until different-witness-replacement is merged?

glozow commented:
> If the plan is to have replacement of different-witness in the future, then why adding this error in the first place here? Do we need this in the meanwhile, until different-witness-replacement is merged?

- We are not adding an error; simply distinguishing between the two cases, returning different error messages, and adding a test for same-txid-different-wtxid.
- The new error message is an improvement upon the current "txn-already-in-mempool" which, in the case of same-txid-different-witness, is inaccurate and confusing for users.
- This is not in conflict with witness replacement; I hope that this speeds it up. :) Merging this first means that PRs that deal with same-txid-different-wtxid don't all conflict with each other.

theStack commented on src/validation.cpp:
```diff
@@ -611,9 +611,13 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
+        // Exact transaction already exists in the mempool.
         return state.Invalid(TxValidationResult::TX_CONFLICT, "txn-already-in-mempool");
+    } else if (m_pool.exists(GenTxid(false, tx.GetHash()))) {
+        // Transaction with the same nonwitness data but different witnes (same txid, different
```
nit: s/nonwitness/non-witness/
Rationale:
```
$ git grep nonwitness | wc -l
0
$ git grep non-witness | wc -l
39
```

glozow replied:
makes sense, will use non-witness from now on!

glozow replied:
Done

theStack reviewed (commented):
Concept ACK 📝 

I think this needs a rebase, as there is a silent merge conflict -- the helper function `ToHex` doesn't exist anymore in master, since #22257 (commit https://github.com/bitcoin/bitcoin/pull/22257/commits/a79396fe5f8f81c78cf84117a87074c6ff6c9d95) has been merged.

jnewbery commented:
Since you're going to rebase, a few v smol suggestions for the test:

<details>
<summary>Diff</summary>

```diff
--- a/test/functional/mempool_accept_wtxid.py
+++ b/test/functional/mempool_accept_wtxid.py
@@ -7,46 +7,42 @@ Test mempool acceptance in case of an already known transaction
 with identical non-witness data different witness.
 """
 
-from test_framework.script import (
-    CScript,
-    OP_0,
-    OP_TRUE,
-    OP_IF,
-    OP_HASH160,
-    OP_EQUAL,
-    OP_ELSE,
-    OP_ENDIF,
-    hash160,
-)
 from test_framework.messages import (
+    COIN,
+    COutPoint,
     CTransaction,
     CTxIn,
     CTxInWitness,
     CTxOut,
-    COutPoint,
     sha256,
-    COIN,
-    ToHex,
 )
-from test_framework.test_framework import BitcoinTestFramework
-from test_framework.util import (
-    assert_equal,
+from test_framework.p2p import P2PTxInvStore
+from test_framework.script import (
+    CScript,
+    OP_0,
+    OP_ELSE,
+    OP_ENDIF,
+    OP_EQUAL,
+    OP_HASH160,
+    OP_IF,
+    OP_TRUE,
+    hash160,
 )
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
 
 class MempoolWtxidTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 1
         self.setup_clean_chain = True
-        self.extra_args = [["-incrementalrelayfee=0"]]
 
     def run_test(self):
         node = self.nodes[0]
 
-        self.log.info('Start with empty mempool, and 200 blocks')
-        privkeys = [node.get_deterministic_priv_key().key]
+        self.log.info("Start with empty mempool")
         # The last 100 coinbase transactions are premature
-        earliest_blockhash = node.generate(101)[0]
-        txid = node.getblock(blockhash=earliest_blockhash, verbosity=2)["tx"][0]["txid"]
+        blockhash = node.generate(101)[0]
+        txid = node.getblock(blockhash=blockhash, verbosity=2)["tx"][0]["txid"]
         assert_equal(node.getmempoolinfo()['size'], 0)
 
         self.log.info("Submit parent with multiple script branches to mempool")
@@ -60,11 +56,12 @@ class MempoolWtxidTest(BitcoinTestFramework):
         parent.vout.append(CTxOut(int(9.99998 * COIN), script_pubkey))
         parent.rehash()
 
+        privkeys = [node.get_deterministic_priv_key().key]
         raw_parent = node.signrawtransactionwithkey(hexstring=parent.serialize().hex(), privkeys=privkeys)['hex']
         parent_txid = node.sendrawtransaction(hexstring=raw_parent, maxfeerate=0)
         node.generate(1)
 
-        # Create a new segwit transaction with witness solving first branch
+        # Create a new transaction with witness solving first branch
         child_witness_script = CScript([OP_TRUE])
         child_witness_program = sha256(child_witness_script)
         child_script_pubkey = CScript([OP_0, child_witness_program])
@@ -77,7 +74,7 @@ class MempoolWtxidTest(BitcoinTestFramework):
         child_one_wtxid = child_one.getwtxid()
         child_one_txid = child_one.rehash()
 
-        # Create another identical segwit transaction with witness solving second branch
+        # Create another identical transaction with witness solving second branch
         child_two = CTransaction()
         child_two.vin.append(CTxIn(COutPoint(int(parent_txid, 16), 0), b""))
         child_two.vout.append(CTxOut(int(9.99996 * COIN), child_script_pubkey))
```

</details>

glozow commented:
Rebased and applied suggestions from https://github.com/bitcoin/bitcoin/pull/22253#issuecomment-875799008

naumenkogs commented:
ACK b7a8cd9963e810264d3b45d0ad15af863965c47a

jnewbery commented:
Code review ACK b7a8cd9963e810264d3b45d0ad15af863965c47a

theStack commented on test/functional/mempool_accept_wtxid.py:
```diff
@@ -0,0 +1,116 @@
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test mempool acceptance in case of an already known transaction
+with identical non-witness data different witness.
```
nit:
```suggestion
with identical non-witness data but different witness.
```

theStack commented on test/functional/mempool_accept_wtxid.py:
```diff
@@ -0,0 +1,116 @@
+            "allowed": False,
+            "reject-reason": "txn-already-in-mempool"
+        }])
+        testres_child_two = node.testmempoolaccept([child_two.serialize().hex()])[0]
+        assert_equal(testres_child_two, {
```
nit, could be simplified (like for `child_one` above):
```suggestion
        assert_equal(node.testmempoolaccept([child_two.serialize().hex()]), [{
```

theStack commented on test/functional/mempool_accept_wtxid.py:
```diff
@@ -0,0 +1,116 @@
+        child_one_txid = child_one.rehash()
+
+        # Create another identical transaction with witness solving second branch
+        child_two = CTransaction()
+        child_two.vin.append(CTxIn(COutPoint(int(parent_txid, 16), 0), b""))
+        child_two.vout.append(CTxOut(int(9.99996 * COIN), child_script_pubkey))
+        child_two.wit.vtxinwit.append(CTxInWitness())
+        child_two.wit.vtxinwit[0].scriptWitness.stack = [b'', witness_script]
```
possible dedup alternative, since the txs only differ in one part:
```suggestion
        child_two = copy.deepcopy(child_one)
        child_two.wit.vtxinwit[0].scriptWitness.stack = [b'', witness_script]
```

theStack reviewed (approved):
Code-review ACK b7a8cd9963e810264d3b45d0ad15af863965c47a

I think the simplest parent witness script to create same-nonwitness-data txs would be `OP_DROP OP_TRUE`, with a different arbitrary witness stack element in each child transaction. OTOH multiple script branches are far more realistic, I guess that's the primary scenario :)

Left some non-blocking nits below only regarding the test, feel free to ignore.

darosior reviewed (approved):
re-utACK b7a8cd9963e810264d3b45d0ad15af863965c47a

jonatack commented on src/validation.cpp:
```diff
@@ -587,9 +587,13 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
+    if (m_pool.exists(GenTxid(true, tx.GetWitnessHash()))) {
+        // Exact transaction already exists in the mempool.
         return state.Invalid(TxValidationResult::TX_CONFLICT, "txn-already-in-mempool");
+    } else if (m_pool.exists(GenTxid(false, tx.GetHash()))) {
```
(One nicety if you pass this way again.)
```diff
-    if (m_pool.exists(GenTxid(true, tx.GetWitnessHash()))) {
+    if (m_pool.exists(GenTxid(/* is_wtxid */ true, tx.GetWitnessHash()))) {
         // Exact transaction already exists in the mempool.
         return state.Invalid(TxValidationResult::TX_CONFLICT, "txn-already-in-mempool");
-    } else if (m_pool.exists(GenTxid(false, tx.GetHash()))) {
+    } else if (m_pool.exists(GenTxid(/* is_wtxid */ false, tx.GetHash()))) {
```
